l1/l1.go

```diff
@@ -75,101 +62,86 @@ func (c *Client) checkChainID(ctx context.Context) error {
 
 	// NOTE: for now we return an error. If we want to support users who fork
 	// Starknet to create a "custom" Starknet network, we will need to log a warning instead.
-	return fmt.Errorf("mismatched L1 and L2 networks: L2 network %s; is the L1 node on the correct network?", c.network)
+	return fmt.Errorf("mismatched L1 and L2 networks:  L1 chain ID: %s, L2 chain ID: %s,"+
+		" is the L1 node on the correct network (see --eth-node)?", wantChainID.Text(felt.Base16), gotChainID.Text(felt.Base16))
 }
 
 func (c *Client) Run(ctx context.Context) error {
-	defer c.l1.Close()
 	if err := c.checkChainID(ctx); err != nil {
 		return err
 	}
 
-	buffer := 128
-
-	logStateUpdateChan := make(chan *contract.StarknetLogStateUpdate, buffer)
-	subUpdates, err := c.subscribeToUpdates(ctx, logStateUpdateChan)
-	if err != nil {
-		return err
-	}
-	defer subUpdates.Unsubscribe()
-
-	headerChan := make(chan *types.Header, buffer)
-	subHeaders, err := c.subscribeToHeaders(ctx, headerChan)
-	if err != nil {
-		return err
-	}
-	defer subHeaders.Unsubscribe()
-
+	ticker := time.NewTicker(c.delay)
+	defer ticker.Stop()
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
-		case err := <-subHeaders.Err():
-			c.log.Warnw("L1 header subscription failed, resubscribing", "error", err)
-			subHeaders.Unsubscribe()
+		case <-ticker.C:
+			latestL1BlockNumber, finalisedL1BlockNumber, latestL1BlockHash, err := c.l1.LatestAndFinalisedHeights(ctx)
+			if err != nil {
+				c.log.Warnw("Failed to fetch finalised and latest L1 block numbers", "err", err)
+				continue
+			}
 
-			subHeaders, err = c.subscribeToHeaders(ctx, headerChan)
+			if c.finalisedHeight > finalisedL1BlockNumber {
+				// TODO how should we handle such catastrophic events?
+				return errReorgFinalisedBlock
+			}
```

Comment:
This could happen when using a RPC provider. Since RPC providers run multiple nodes and put them behind a load balancer, our requests could be handled by different nodes. So getting a new finalized height that is lower than our c.finalisedHeight is not necessarily a catastrophic failure, it might just be that one of the nodes behind the load balancer could be a bit behind.

So I think we should simply ignore it instead of shutting the service down.

Reply:
Thanks. Changed code to use event subscription, these lines are no longer present.